### PR TITLE
Pol 408 implement a policy to create an incident when update is required in the json for instance types

### PR DESCRIPTION
**Pull request by SheikSena**
### Description

The Policy finds instance/machine types across the AWS, Azure, Google cloud vendors and compares them with the existing machine/instance types in our data file. Incident will be created if any new machine/instance type is found, containing missing machine/instance type name and vendor details.

### Contribution Check List

- [x] All tests pass.
- [x] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable
- [x] New functionality has been documented in CHANGELOG.MD

Policy Incident Link: https://governance.rightscale.com/org/1105/projects/60073/policy-incidents/5fd9d5d467016a0001e898e6

image

### CODE REVIEW

File: compliance/instance_types/list_new_instance_types.pt [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+    auth $auth_google
+    pagination $google_pagination
+    host "compute.googleapis.com"
+    path join(["/compute/v1/projects/",val(iter_item, "projectId"),"/zones"])
+  end
+  result do
+    encoding "json"
+    collect jmes_path(response,"items[*]") do
+      field "zone", jmes_path(col_item, "name")
+	  field "projectId", val(iter_item, "projectId")
```
Comment (nshivappa) [inline compliance/instance_types/list_new_instance_types.pt:218]: Check the line indentation.

File: compliance/instance_types/list_new_instance_types.pt [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+      if(aws_instance_type_check_in_github == null){
+        new_instance_types_found.push({
+          "instanceType": value.instanceType,
+          "cloudProvider": "AWS"
+        })
+      }
+    })
+
+    // Finding new instance types in Azure
+    var  azure_unique_instance_types =     _.uniq(_.map(azure_instance_types, function(d){return d.instanceType}));
```
Comment (nshivappa) [inline compliance/instance_types/list_new_instance_types.pt:309]:
```suggestion
    var  azure_unique_instance_types = _.uniq(_.map(azure_instance_types, function(d){return d.instanceType}));
```

File: compliance/instance_types/list_new_instance_types.pt [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+    _.each(google_SQL_tyres_list, function(sql){
+      if(machines_list_obj[sql.name]==null){
+	      //filtering out 1st gen db tiers
+	      if(sql.name.indexOf("D")==0){
+        }
+		    else{
+		      machines_list_obj[sql.name]=sql;
+		    }
+	    }
+	  })
```
Comment (nshivappa) [inline compliance/instance_types/list_new_instance_types.pt:339]: Check the line indentation.

File: compliance/instance_types/list_new_instance_types.pt [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+	      if( machine == "n1-megamem-96" || machine == "n1-ultramem-160" || machine == "n1-ultramem-40" || machine == "n1-ultramem-80"){
+		      continue;
+		    }
+        console.log(machine);
+        new_instance_types_found.push({
+          "instanceType": machine,
+          "cloudProvider": "Google"
+        })
+	    }
+	  }
```
Comment (nshivappa) [inline compliance/instance_types/list_new_instance_types.pt:358]: Check the line indentation.

File: compliance/instance_types/list_new_instance_types.pt [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+end
+
+datasource "ds_azure_instance_types" do
+  iterate $ds_azure_locations
+  request do
+    auth $azure_auth
+    host "management.azure.com"
+    path join(["/subscriptions/", val(iter_item,"subscriptionId"), "/providers/Microsoft.Compute/locations/", val(iter_item,"locationName"),"/vmSizes"])
+    query "api-version","2020-06-01"
+    ignore_status [400]
```
Comment (nshivappa) [inline compliance/instance_types/list_new_instance_types.pt:174]: Any particular reason to ignore the 400 status code

**Review by nshivappa — COMMENTED** [review]
Left few comments

File: compliance/instance_types/README.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+## What it does
+
+The Policy finds instance/machine types across the AWS, Azure, Google cloud vendors and compares them with the existing machine/instance types in our data file. Incident will be created if any new machine/instance type if found, containing missing machine/instance type name and vendor details.
+
+## Input Parameters
+
+This policy has the following input parameters required when launching the policy.
+
+- *Email addresses* - A list of email addresses to notify
```
[the remark this answers is not in the document]
Comment (SheikSena) [reply compliance/instance_types/README.md]: Done

File: compliance/instance_types/README.md [hunk 7]
```diff
@@ -0,0 +1,62 @@
+# New Instance Types
+
+## What it does
+
+The Policy finds instance/machine types across the AWS, Azure, Google cloud vendors and compares them with the existing machine/instance types in our data file. Incident will be created if any new machine/instance type if found, containing missing machine/instance type name and vendor details.
+
+## Input Parameters
+
+This policy has the following input parameters required when launching the policy.
```
[the remark this answers is not in the document]
Comment (SheikSena) [reply compliance/instance_types/README.md]: Done

File: compliance/instance_types/README.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+- *Email addresses* - A list of email addresses to notify
+
+## Policy Actions
+
+The following policy actions are taken on any resources found to be out of compliance.
```
[the remark this answers is not in the document]
Comment (SheikSena) [reply compliance/instance_types/README.md]: Done

File: compliance/instance_types/README.md [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    "Statement":[
+        {
+        "Effect":"Allow",
+        "Action":[
+            "ec2:DescribeInstances"
+        ],
+        "Resource":"*"
+        }
+    ]
+    }
```
[the remark this answers is not in the document]
Comment (SheikSena) [reply compliance/instance_types/README.md]: Done

File: compliance/instance_types/list_new_instance_types.pt [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+  policy_set: "New Instance Types"
+)
+
+###############################################################################
+# User inputs
+###############################################################################
+
+parameter "param_email" do
+  type "list"
+  label "Email addresses"
```
[the remark this answers is not in the document]
Comment (SheikSena) [reply compliance/instance_types/list_new_instance_types.pt]: Done

File: compliance/instance_types/list_new_instance_types.pt [hunk 11]
```diff
@@ -0,0 +1,395 @@
...
+  description "A list of email addresses to notify."
```
[the remark this answers is not in the document]
Comment (SheikSena) [reply compliance/instance_types/list_new_instance_types.pt]: Done

File: compliance/instance_types/list_new_instance_types.pt (hunk 1 above)
Comment (SheikSena) [reply compliance/instance_types/list_new_instance_types.pt]: Done

File: compliance/instance_types/list_new_instance_types.pt (hunk 2 above)
Comment (SheikSena) [reply compliance/instance_types/list_new_instance_types.pt]: Done

File: compliance/instance_types/list_new_instance_types.pt (hunk 3 above)
Comment (SheikSena) [reply compliance/instance_types/list_new_instance_types.pt:338]: The format is correct in the pt file but in Github, it is not showing the correct format.

File: compliance/instance_types/list_new_instance_types.pt (hunk 4 above)
Comment (SheikSena) [reply compliance/instance_types/list_new_instance_types.pt]: The format is correct in the pt file but in Github, it is not showing the correct format.

**Comment by nshivappa** [comment]
- Please look into the flexera-ci ERROR
- Can you check if the instance type in AWS is the same across all the regions
- And please check if it is possible to display up and down values in incident when new instance types are found.

File: compliance/instance_types/list_new_instance_types.pt (hunk 3 above)
Comment (nshivappa) [reply compliance/instance_types/list_new_instance_types.pt:338]: See if it contains tab space.

File: compliance/instance_types/list_new_instance_types.pt (hunk 4 above)
Comment (nshivappa) [reply compliance/instance_types/list_new_instance_types.pt]: See if it contains tab space.

File: compliance/instance_types/list_new_instance_types.pt (hunk 3 above)
Comment (SheikSena) [reply compliance/instance_types/list_new_instance_types.pt]: Done

File: compliance/instance_types/list_new_instance_types.pt (hunk 5 above)
Comment (SheikSena) [reply compliance/instance_types/list_new_instance_types.pt:174]: We were getting 400 error so, I have added ignore status 400.

Error Message:
{"error":{"code":"NoRegisteredProviderFound","message":"No registered resource provider found for location 'centralusstage' and API version '2020-06-01' for type 'locations/vmSizes'. The supported api-versions are '2015-05-01-preview, 2015-06-15, 2016-03-30, 2016-04-30-preview, 2016-08-30, 2017-03-30, 2017-12-01, 2018-04-01, 2018-06-01, 2018-10-01, 2019-03-01, 2019-07-01, 2019-12-01, 2020-06-01, 2020-12-01'. The supported locations are 'eastus, eastus2, westus, centralus, northcentralus, southcentralus, northeurope, westeurope, eastasia, southeastasia, japaneast, japanwest, australiaeast, australiasoutheast, australiacentral, brazilsouth, southindia, centralindia, westindia, canadacentral, canadaeast, westus2, westcentralus, uksouth, ukwest, koreacentral, koreasouth, francecentral, southafricanorth, uaenorth, switzerlandnorth, germanywestcentral, norwayeast'."}}

**Comment by SheikSena** [comment]
@nshivappa, I tried to display the Up and Down value for instance type in the incident but it is not feasible because the instance type names pattern is not similar, so it is very challenging to get Up and Down value for an instance type.

**Review by rshade — APPROVED** [review]
LGTM

**Pull request merged**